Rules/UseUsingScopeModifierInNewRunspaces.cs

```diff
@@ -0,0 +1,375 @@
+            {
+                yield break;
+            }
+
+            foreach (var scriptBlockAst in scriptBlockAsts)
+            {
+                var sbAst = scriptBlockAst as ScriptBlockAst;
+
+                foreach (var diagnosticRecord in AnalyzeScriptBlockAst(sbAst, fileName))
+                {
+                    yield return diagnosticRecord;
+                }
+            }
+        }
+
+        /// <summary>
+        /// GetName: Retrieves the name of this rule.
+        /// </summary>
+        /// <returns>The name of this rule</returns>
+        public string GetName()
+        {
+            return string.Format(CultureInfo.CurrentCulture, Strings.NameSpaceFormat, GetSourceName(),
+                Strings.UseUsingScopeModifierInNewRunspacesName);
+        }
+
+        /// <summary>
+        /// GetCommonName: Retrieves the common name of this rule.
+        /// </summary>
+        /// <returns>The common name of this rule</returns>
+        public string GetCommonName()
+        {
+            return string.Format(CultureInfo.CurrentCulture, Strings.UseUsingScopeModifierInNewRunspacesCommonName);
+        }
+
+        /// <summary>
+        /// GetDescription: Retrieves the description of this rule.
+        /// </summary>
+        /// <returns>The description of this rule</returns>
+        public string GetDescription()
+        {
+            return string.Format(CultureInfo.CurrentCulture, Strings.UseUsingScopeModifierInNewRunspacesDescription);
+        }
+
+        /// <summary>
+        /// GetSourceType: Retrieves the type of the rule: builtin, managed or module.
+        /// </summary>
+        public SourceType GetSourceType()
+        {
+            return SourceType.Builtin;
+        }
+
+        /// <summary>
+        /// GetSeverity: Retrieves the severity of the rule: error, warning of information.
+        /// </summary>
+        /// <returns></returns>
+        public RuleSeverity GetSeverity()
+        {
+            return RuleSeverity.Warning;
+        }
+
+        /// <summary>
+        /// GetSourceName: Retrieves the module/assembly name the rule is from.
+        /// </summary>
+        public string GetSourceName()
+        {
+            return string.Format(CultureInfo.CurrentCulture, Strings.SourceName);
+        }
+
+        /// <summary>
+        /// Checks if a variable is initialized and referenced in either its assignment or children scopes
+        /// </summary>
+        /// <param name="scriptBlockAst">Ast of type ScriptBlock</param>
+        /// <param name="fileName">Name of file containing the ast</param>
+        /// <returns>An enumerable containing diagnostic records</returns>
+        private IEnumerable<DiagnosticRecord> AnalyzeScriptBlockAst(ScriptBlockAst scriptBlockAst, string fileName)
+        {
+            var astsToProcess = new List<Ast>();
+            var nonAssignedNonUsingVars = new List<VariableExpressionAst>();
+
+            astsToProcess.AddRange(

```

Comment:
You'd probably be much better off writing an `AstVisitor2` to accomplish this, since that allows you to simply override some methods to visit the ASTs you're looking for.

[Here](https://github.com/PowerShell/PSScriptAnalyzer/blob/f389aa42166a8fdbf293432eb0c8f42aa7c11508/Rules/CompatibilityRules/UseCompatibleSyntax.cs#L152-L376)'s an example

Reply:
do you know in which PS version AstVisitor2 was introduced as PSSA still has to support PSv3 unfortunately and therefore we might need conditional compilation for that

Reply:
on it